presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java

```diff
@@ -93,6 +97,59 @@ public JoinNode(@JsonProperty("id") PlanNodeId id,
         checkArgument(!isCrossJoin() || inputSymbols.equals(outputSymbols), "Cross join does not support output symbols pruning or reordering");
     }
 
+    public JoinNode flip()
+    {
+        return new JoinNode(
+                getId(),
+                flipType(type),
+                right,
+                left,
+                flipJoinCriteria(criteria),
+                flipOutputSymbols(getOutputSymbols(), left, right),
+                filter,
+                rightHashSymbol,
+                leftHashSymbol,
+                distributionType);
+    }
+
+    private static Type flipType(Type type)
+    {
+        switch (type) {
+            case INNER:
+                return INNER;
+            case FULL:
+                return FULL;
+            case LEFT:
+                return RIGHT;
+            case RIGHT:
+                return LEFT;
+            default:
+                throw new IllegalStateException("No inverse defined for join type: " + type);
+        }
+    }
+
+    private static List<EquiJoinClause> flipJoinCriteria(List<EquiJoinClause> joinCriteria)
+    {
+        return joinCriteria
+                .stream()
+                .map(equiJoinClause -> new EquiJoinClause(equiJoinClause.getRight(), equiJoinClause.getLeft()))
+                .collect(toImmutableList());
+    }
+
+    private static List<Symbol> flipOutputSymbols(List<Symbol> outputSymbols, PlanNode left, PlanNode right)
+    {
+        List<Symbol> leftSymbols = outputSymbols.stream()
+                .filter(symbol -> left.getOutputSymbols().contains(symbol))
+                .collect(Collectors.toList());
+        List<Symbol> rightSymbols = outputSymbols.stream()
+                .filter(symbol -> right.getOutputSymbols().contains(symbol))
+                .collect(Collectors.toList());
+        return ImmutableList.<Symbol>builder()
+                .addAll(rightSymbols)
+                .addAll(leftSymbols)
```

Comment:
it's necessary for cross joins because cross joins don't support output symbol reordering.